phd-tests/tests/src/disk.rs

```diff
@@ -66,3 +142,73 @@ async fn in_memory_backend_smoke_test(ctx: &Framework) {
     let cat = vm.run_shell_command("cat /phd/hello_oxide.txt").await?;
     assert_eq!(cat, HELLO_MSG);
 }
+
+#[phd_testcase]
+async fn in_memory_backend_migration_test(ctx: &Framework) {
+    // This test verifies that live-migrating a disk with an in-memory backend
+    // copies the backend's data from the source to the target. To do that, it
+    // uses `dd` flags to (try to) force writes to be synchronized to disk
+    // immediately and to force reads to bypass the guest OS's caches.
+    //
+    // Dynamic tracing of block backend activity shows that Alpine guests don't
+    // reliably read from underlying storage even when `iflag=direct` is used.
+    // This can cause this test to pass incorrectly. Unless and until a more
+    // reliable way to bypass caches is found, skip this test on Alpine.
+    if let GuestOsKind::Alpine = ctx.default_guest_os_kind().await? {
+        phd_skip!("iflag=direct doesn't work as required on Alpine");
+    }
+
+    // A blank disk is fine for this test: the rest of the test will address the
+    // disk device directly instead of assuming it has a file system. This works
+    // around #824 for Windows guests (which may not recognize the FAT
+    // filesystems PHD produces).
+    let (vm, device_path) = launch_vm_and_find_in_memory_disk(
+        ctx,
+        "in_memory_backend_migration_test_source",
+        DiskSource::Blank(16 * 1024),
+        false,
+    )
+    .await?;
+
+    // Scribble random data into the first kilobyte of the data disk. Use
+    // `oflag=sync` to force the guest OS to ensure this data is actually
+    // persisted to the device (and not just held in an in-memory cache).
+    vm.run_shell_command(&format!(
+        "dd if=/dev/random of={device_path} oflag=sync bs=1K count=1"
+    ))
+    .await?;
+
+    // Read the scribbled data out to a file on the main OS disk.
+    vm.run_shell_command(&format!(
+        "dd if={device_path} of=/tmp/before iflag=direct bs=1K"
+    ))
+    .await?;
```

Comment:
idle thought, might be Alpine-inclusive while we don't know what's up with `dd iflag=direct`: `sync && echo 3 /proc/sys/vm/drop_caches` is (if memory serves) synchronous and will also ensure that written data has been flushed out. caches should then also be empty for reads later on.

but `O_DIRECT` is more obvious at time of use of course. so really more a "if you didn't think of it, might be interesting" - what's here seems good.